csharp/src/Google.Protobuf/JsonTokenizer.cs

```diff
@@ -66,6 +73,8 @@ internal JsonTokenizer(TextReader reader)
             containerStack.Push(ContainerType.Document);
         }
 
+        // TODO: Why do we allow a different token to be pushed back? It might be better to always remember the previous
```

Comment:
I agree here, a bit of refactoring would probably get rid of  the need to push to token back and of some of the objectDepth-- and objectDepth++ adjustments for buffered tokens.

For now I think it's totally fine to go with what we currently have.